fabric: Add mr_count field to domain attribute

shefty opened the pull request:
Fixes #3084

Allow the provider to indicate the number of memory registrations
that it can support on a domain.  This also allows the app a way
to indicate how many memory regions it may need.  Providers can
use this value to optimize caching of registrations.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

shefty commented:
@j-xiong @a-ilango @pkcoff @hppritcha @jswaro @hoopoepg @gladkovdmitry17 - This adds a new domain attribute.  Please see updates to providers to set it and let me know if those should change and to what.  Thanks.

jswaro commented on prov/gni/src/gnix_fabric.c:
```diff
@@ -253,6 +253,7 @@ static struct fi_info *_gnix_allocinfo(void)
 	gnix_info->domain_attr->mr_iov_limit = 1;
 	gnix_info->domain_attr->caps = GNIX_DOM_CAPS;
 	gnix_info->domain_attr->mode = 0;
+	gnix_info->domain_attr->mr_cnt = gnix_default_user_registration_limit;
```
Please set the value to (65536) for now. Basic registration (which is the current default for the GNI provider) has a much larger pool of registrations than the default specified in that variable (which was/is intended for use with scalable memory registration.

shefty replied:
Will change - thanks

j-xiong commented on prov/psm/src/psmx_init.c:
```diff
@@ -558,6 +558,7 @@ static int psmx_getinfo(uint32_t version, const char *node, const char *service,
 	psmx_info->domain_attr->mr_iov_limit = 65535;
 	psmx_info->domain_attr->caps = PSMX_DOM_CAPS;
 	psmx_info->domain_attr->mode = 0;
+	psmx_info->domain_attr->mr_cnt = 10000;
```
Number of MRs in psm and psm2 is only limited by the key, which is 64 bits. So -1ULL would work here. For practical purpose, any number that is large enough should be fine, but value like 65535 or 2097151 is more preferable than 10000.

shefty replied:
I picked 10000, since I saw that in the registration code as a number of tries.

j-xiong replied:
But the try starts from the key previously used, not from 0.

shefty replied:
got it - will update to -1ULL

jswaro replied:
Is -1 analogous to 'unlimited' or something similar in the context of this new field?

gladkovdmitry17 replied:
@jswaro `(size_t)-1ULL` is max value of `size_t ` datatype.

j-xiong replied:
That's just the maximum value of the mr key, and MRs can't share the same key .... in practical it's unlimited, the value is not treated differently.

jswaro replied:
> @jswaro (size_t)-1ULL is max value of size_t datatype.

That wasn't what I was asking. From the perspective of libfabric as an API, will we be considering -1 to mean 'unlimited'?

If I'm writing an application and I see that value, does that functionally mean I could register as many regions as I wanted (or more to the point, that the provider doesn't have a limit that it can easily express or nor does it have an informative limit)? My question is more directed toward how applications should interpret -1.

shefty replied:
To be clear, the fields are unsigned, so the value is better represented as SIZE_MAX, not -1. I would interpret such a value to mean that the provider limit is tied to memory constraints.  This applies to any of the size_t attribute fields.

jswaro replied:
Thanks.

gladkovdmitry17 replied:
> @jswaro (size_t)-1ULL is max value of size_t datatype.
> That wasn't what I was asking. From the perspective of libfabric as an API, will we be considering -1 to mean 'unlimited'?

Sorry, I got your question incorrectly.

jswaro replied:
Not a problem. I wasn't being very clear I suppose.

shefty commented:
Updated with changes from @gladkovdmitry17 and based on feedback comments.  I changed my mind and picked 65535 for the psm limits, since -1ULL is not technically feasible given that each registration must allocate memory and I didn't want something like (size_t) -1ULL / sizeof(struct psmx_mr).

On a separate note, with this new field, the util code may be able to provide optimized key caches based on the application mr_cnt hints.

gladkovdmitry17 commented:
> Updated with changes from @gladkovdmitry17 and based on feedback comments.

Thanks. Seems new unstable test is discovered (`fi_poll`) in Appveyor CI testing on Windows. I'll check it tomorrow.
